designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+## Constraint Block
+
+```stan
+constraints {
+ upper_bound {
+   constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+```
+
+The new `constraints` block will need to be added to the compiler's parser, validator, mir, and c++ mir. The parser will transpile each of the constraints to c++ functions which start with the name of the constraint. For the example above the signatures would be
+
+```c++
+upper_bound_constrain__(...)
+upper_bound_unconstrain__(...)
+upper_bound_validate__(...)
+```
+
+The only function of the three listed above that has rules that differ from standard function generation is the `*_constrain` function which has an additional `jacobian` argument and `Jacobian` template parameter. The jacobian is kept behind an if statement that checks the compile time value `Jacobian` before incrementing the jacobian. Besides the jacobian argument the rest of the function will parse exactly like a standard function.
+
+```c++
+template <bool Jacobian, typename T1, typename T2, typename TJacobian>
+return_type_t<T1, T2> upper_bound_constrain(const T1& x, const T2& upper_bound,
+  TJacobian& jacobian) {
+ if (Jacobian) {
+   // If x is a vector we can
+   //  wrap the right hand side in a sum()
+   jacobian += x;
+ }
+ return subtract(upper_bound, exp(x));
+}
+```
+
+The `_unconstrain` function will not be used directly by the user and instead will be used internally in the Stan model class when we need to go from the constrained to unconstrained space. The code generation for these will be the same as a standard function.
+
+## Transformed Parameters
+
+Inside of transformed parameters we expose `jacobian` like in the functions section listed below. Any jacobian accumulation will be wrapped in an `if (Jacobian)` so that the Jacobian increment only happens when requested.
+
+```stan
+parameters {
+ real b_raw;
+}
+transformed parameters {
+ // (2) Transform and accumulates like stan math directly
+ real b = exp(b_raw) + lower_bound;
+ // Underneath the hood, only actually calculated if
+ // Jacobian bool template parameter is set to true
+ jacobian += b_raw;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Making a new keyword `jacobian` will almost surely break current user code.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+
+The other alternative is to not have a new block and have users define `*_constrain`, `*_unconstrain`, and `*_validate` functions in the `functions` block. This might be easier to implement, but it is not as nice looking as a new constraints block.
```

Comment:
Be more specific.  It doesn't look nice because it's not localized and encapsulated.  Just say that.

Reply:
👍